# Subway Data Analysis

## Introduction

O sistema de ônibus e trens de Nova Iorque - o Metro Transit Authority - [fornece seus dados para download](http://web.mta.info/developers/developer-data-terms.html#data) através de  arquivos CSV. Dentre as informações disponíveis estão os **registros semanais de dados das catracas do metrô**. 


Estes registros contém contagens cumulativas das entradas e saídas, normalmente agrupadas em períodos de 4 horas, com dados adicionais que permitem identificar a estação e catraca específica correspondente a cada linha do arquivo. Neste projeto iremos utilizar um desses registros, mas não precisa baixar nada agora! O primeiro exercício será escrever um código Python para fazer isso por você :-)




# Sobre este projeto

Neste projeto você irá aplicar todos os conhecimentos adquiridos neste primeiro mês de curso. Iremos praticar tarefas básicas de aquisição, limpeza de dados e nesse processo iremos descobrir coisas essenciais sobre os dados utilizando o que foi aprendido no curso de estatística. 

O objetivo deste projeto é explorar a relação entre os dados das catracas do metro de Nova Iorque e o clima no dia da coleta. Para isso, além dos dados do metrô, precisaremos os dados de clima da cidade de Nova Iorque. 

Os principais pontos que serão verificados neste trabalho:

- Coleta de dados da internet
- Utilização de estatística para análise de dados
- Manipulação de dados e criação de gráficos simples com o `Pandas`

*Como conseguir ajuda*: Sugerimos que tente os seguintes canais, nas seguintes ordens:

| Tipo de dúvida\Canais         	| Google 	| Fórum 	| Slack 	| Email 	|
|-------------------------------	|--------	|-------	|-------	|-------	|
| Programação Pyhon e Pandas    	| 1      	| 2     	| 3     	|       	|
| Requisitos do projeto         	|        	| 1     	| 2     	| 3     	|
| Partes específicas do Projeto 	|        	| 1     	| 2     	| 3     	|

Os endereços dos canais são:

- Fórum: https://discussions.udacity.com/c/ndfdsi-project
- Slack: [udacity-br.slack.com](https://udacity-br.slack.com/messages/C5MT6E3E1)
- Email: data-suporte@udacity.com

**Espera-se que o estudante entregue este relatório com:**

- Todos os exercícios feitos, com atenção especial para os trechos de código a completar (sinalizados com `# your code here`), pois eles são essenciais para que o código rode corretamente
- O arquivo ipynb exportado como HTML

Para entregar este projeto envie este `.ipynb` preenchido e o HTML, zipados, na página correspondente da sala de aula.

# Sobre o dataset

Descrição das colunas
<pre>
C/A,UNIT,SCP,STATION,LINENAME,DIVISION,DATE,TIME,DESC,ENTRIES,EXITS
  
C/A     = Agrupamento de catracas de que a catraca faz parte (_Control Area_)
UNIT    = Cabine de controle associada á  estação onde a catraca se encontra (_Remote Unit for a station_)
SCP     = Endereço especí­fico da catraca (_Subunit Channel Position_)
STATION = Nome da estação onde a catraca se encontra
LINENAME= Código representando todas linhas que passam na estação*
DIVISION= Código representando a concessionária original da linha, antes da prefeitura assumir a gestão   
DATE    = Representa a data (no formato MM-DD-YY) do evento de auditoria agendado
TIME    = Representa o horário (hh:mm:ss) do evento de auditoria agendado
DESc    = Descreve o tipo de evento de auditoria registrado:
          1. "REGULAR" representando um evento de auditoria padrão, em que a contagem é feita a cada 4 horas
          2. "RECOVR AUD" significa que o valor especí­fico estava perdido, mas foi recuperado posteriormente 
          3. Diversos códigos sinalizam situações em que auditorias são mais frequentes devido a atividades
            de planejamento ou solução de problemas. 
ENTRIES = A contagem cumulativa de entradas associadas á catraca desde o último registro
EXITS   = A contagem cumulativa de saídas associadas á catraca desde o último registro

*  Normalmente as linhas são representadas por um caractere. LINENAME 456NQR significa que os trens 4, 5, 6, N, Q e R passam pela estação.
</pre>

# Lembretes

Antes de começarmos, alguns lembretes devem ter em mente ao usar os notebooks iPython:

- Lembre-se de que você pode ver do lado esquerdo de uma célula de código quando foi executado pela última vez se houver um número dentro das chaves.
- Quando você inicia uma nova sessão do notebook, certifique-se de executar todas as células até o ponto em que você deixou a última vez. Mesmo que a saída ainda seja visível a partir de quando você executou as células em sua sessão anterior, o kernel começa em um estado novo, então você precisará recarregar os dados, etc. em uma nova sessão.
- O ponto anterior é útil para ter em mente se suas respostas não correspondem ao que é esperado nos questionários da aula. Tente recarregar os dados e execute todas as etapas de processamento um a um para garantir que você esteja trabalhando com as mesmas variáveis e dados que estão em cada fase do questionário.

## Seção 1 - Coleta de Dados

### *Exercicio 1.1*

Mãos a obra!! Agora é sua vez de coletar os dados. Escreva abaixo um código python que acesse o link http://web.mta.info/developers/turnstile.html e baixe os arquivos do mês de junho de 2017. O arquivo deverá ser salvo com o nome turnstile_100617.txt onde 10/06/17 é a data do arquivo.

Abaixo seguem alguns comandos que poderão te ajudar:

Utilize a biblioteca **urllib** para abrir e resgatar uma página da web. Utilize o comando abaixo onde **url** será o caminho da página da web onde se encontra o arquivo:

```python
u = urllib.urlopen(url)
html = u.read()
```

Utilize a biblioteca **BeautifulSoup** para procurar na página pelo link do arquivo que deseja baixar. Utilize o comando abaixo para criar o seu objeto *soup* e procurar por todas as tags 'a'no documento:
 
 
```python
soup = BeautifulSoup(html, "html.parser")
links = soup.find_all('a')
```

Uma dica para baixar apenas os arquivos do mês de junho é verificar a data no nome do arquivo. Por exemplo, para baixar o arquivo do dia 17/06/2017 verifique se o link termina com *"turnstile_170610.txt"*. Se não fizer isso você baixará todos os arquivos da página. Para fazer isso utilize o comando conforme abaixo:

```python
if '1706' in link.get('href'):
```

E a dica final é utilizar o comando abaixo para fazer o download do arquivo txt:

```python
urllib.urlretrieve(link_do_arquivo, filename)
```

Lembre-se, primeiro, carregue todos os pacotes e funções que você estará usando em sua análise.

In [5]:
import urllib
from bs4 import BeautifulSoup

url_base= 'http://web.mta.info/developers/'
u = urllib.urlopen(url_base + 'turnstile.html')
html = u.read()

soup = BeautifulSoup(html, "html.parser")
links = soup.find_all('a')

filenames = []

for link in links:
    url = link.get('href')
    if url == None:
        continue
    
    if '1706' in url:
        filename = url.rsplit("/", 1)[-1]
        filenames.append(filename)
        urllib.urlretrieve(url_base + url, filename)

### *Exercicio 1.2*

Escreva uma função que pegue a lista de nomes dos arquivos que você baixou no exercicio 1.1 e consolide-os em um único arquivo. Deve existir apenas uma linha de cabeçalho no arquivo de saida. 

Por exemplo, se o arquivo_1 tiver:
linha 1...
linha 2...

e o outro arquivo, arquivo_2 tiver:
linha 3...
linha 4...
linha 5...

Devemos combinar o arquivo_1 com arquivo_2 em um arquivo mestre conforme abaixo:

'C/A, UNIT, SCP, DATEn, TIMEn, DESCn, ENTRIESn, EXITSn'
linha 1...
linha 2...
linha 3...
linha 4...
linha 5...

**OBS:** Note que algumas colunas foram descartadas!

In [6]:
def create_master_turnstile_file(filenames, output_file):
    with open(output_file, 'w') as master_file:
        master_file.write('C/A,UNIT,SCP,DATEn,TIMEn,DESCn,ENTRIESn,EXITSn\n')
        for filename in filenames:
            with open(filename, 'r') as f:
                f.readline()
                print f.readline()
                master_file.write(f.read())
            
output_file = 'turnstile_100617.txt'
create_master_turnstile_file(filenames, output_file)

A002,R051,02-00-00,59 ST,NQR456W,BMT,06/17/2017,00:00:00,REGULAR,0006224816,0002107317                                            

A002,R051,02-00-00,59 ST,NQR456W,BMT,06/10/2017,00:00:00,REGULAR,0006215258,0002104297                                            

A002,R051,02-00-00,59 ST,NQR456W,BMT,06/03/2017,00:00:00,REGULAR,0006203652,0002101226                                            

A002,R051,02-00-00,59 ST,NQR456W,BMT,05/27/2017,00:00:00,REGULAR,0006195217,0002098317                                            



### *Exercicio 1.3*

Neste exercício, escreva um função que leia o master_file criado no exercicio anterior e carregue-o em um pandas dataframe. Esta função deve filtrar para que o dataframe possua apenas linhas onde a coluna "DESCn" possua o valor "Regular".

Por exemplo, se o data frame do pandas estiver conforme abaixo:
    
    ,C/A,UNIT,SCP,DATEn,TIMEn,DESCn,ENTRIESn,EXITSn
    0,A002,R051,02-00-00,05-01-11,00:00:00,REGULAR,3144312,1088151
    1,A002,R051,02-00-00,05-01-11,04:00:00,DOOR,3144335,1088159
    2,A002,R051,02-00-00,05-01-11,08:00:00,REGULAR,3144353,1088177
    3,A002,R051,02-00-00,05-01-11,12:00:00,DOOR,3144424,1088231

O dataframe deverá ficar conforme abaixo depois de filtrar apenas as linhas onde a coluna DESCn possua o valor REGULAR:

    0,A002,R051,02-00-00,05-01-11,00:00:00,REGULAR,3144312,1088151
    2,A002,R051,02-00-00,05-01-11,08:00:00,REGULAR,3144353,1088177


In [7]:
import pandas

def filter_by_regular(filename):
    
    turnstile_data = pandas.read_csv(filename, sep=',')
    turnstile_data = turnstile_data[turnstile_data['DESCn'] == 'REGULAR']
    return turnstile_data

df = filter_by_regular(output_file)

### *Exercicio 1.4*


Os dados do metrô de NY possui dados cumulativos de entradas e saidas por linha. Assuma que você possui um dataframe chamado df que contém apenas linhas para uma catraca em particular (unico SCP, C/A, e UNIT). A função abaixo deve alterar essas entradas cumulativas para a contagem de entradas desde a última leitura (entradas desde a última linha do dataframe).

Mais especificamente, você deverá fazer duas coisas:

1. Criar uma nova coluna chamada ENTRIESn_hourly

2. Inserir nessa coluna a diferença entre ENTRIESn da linha atual e a da linha anterior. Se a linha possuir alguma NAN, preencha/substitua por 1.

Dica: as funções do pandas shift() e fillna() pode ser úteis nesse exercicio.

Abaixo tem um exemplo de como seu dataframe deve ficar ao final desse exercicio:

           C/A  UNIT       SCP     DATEn     TIMEn    DESCn  ENTRIESn    EXITSn  ENTRIESn_hourly
    0     A002  R051  02-00-00  05-01-11  00:00:00  REGULAR   3144312   1088151                1
    1     A002  R051  02-00-00  05-01-11  04:00:00  REGULAR   3144335   1088159               23
    2     A002  R051  02-00-00  05-01-11  08:00:00  REGULAR   3144353   1088177               18
    3     A002  R051  02-00-00  05-01-11  12:00:00  REGULAR   3144424   1088231               71
    4     A002  R051  02-00-00  05-01-11  16:00:00  REGULAR   3144594   1088275              170
    5     A002  R051  02-00-00  05-01-11  20:00:00  REGULAR   3144808   1088317              214
    6     A002  R051  02-00-00  05-02-11  00:00:00  REGULAR   3144895   1088328               87
    7     A002  R051  02-00-00  05-02-11  04:00:00  REGULAR   3144905   1088331               10
    8     A002  R051  02-00-00  05-02-11  08:00:00  REGULAR   3144941   1088420               36
    9     A002  R051  02-00-00  05-02-11  12:00:00  REGULAR   3145094   1088753              153
    10    A002  R051  02-00-00  05-02-11  16:00:00  REGULAR   3145337   1088823              243

In [8]:
import pandas

def get_hourly_entries(df):
    df['ENTRIESn_hourly'] = (df['ENTRIESn'] - df['ENTRIESn'].shift(1)).fillna(1).apply(int)
    return df

get_hourly_entries(df)

C/A     UNIT  SCP       DATEn     TIMEn  \
A002  R051 02-00-00          59 ST  NQR456W  BMT  06/17/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/22/2017  00:00:00   
...                            ...      ...  ...         ...       ...   
TRAM2 R469 00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  17:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  21:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  17:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  21:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  17:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  21:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT 

### *Exercicio 1.5*

Faça o mesmo do exercicio anterior mas agora considerando as saidas, coluna EXITSn.
Para isso crie uma coluna chamada de EXITSn_hourly e insira a diferença entre a coluna EXITSn da linha atual versus a linha anterior. Se tiver algum NaN, preencha/substitua por 0.



In [9]:
import pandas

def get_hourly_exits(df):
    df['EXITSn_hourly'] = (df['EXITSn'] - df['EXITSn'].shift(1)).fillna(0).apply(int)
    return df
get_hourly_exits(df)

C/A     UNIT  SCP       DATEn     TIMEn  \
A002  R051 02-00-00          59 ST  NQR456W  BMT  06/17/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/17/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/18/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/19/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/20/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  00:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  04:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  08:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  12:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  16:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/21/2017  20:00:00   
           02-00-00          59 ST  NQR456W  BMT  06/22/2017  00:00:00   
...                            ...      ...  ...         ...       ...   
TRAM2 R469 00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  17:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/29/2017  21:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  17:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/30/2017  21:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  17:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  05/31/2017  21:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  01:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  05:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  09:00:00   
           00-05-01  RIT-ROOSEVELT        R  RIT  06/01/2017  13:00:00   
           00-05-01  RIT-ROOSEVELT 

### *Exercicio 1.6*

Dado uma variável de entrada que representa o tempo no formato de: "00:00:00" (hora: minutos: segundos)

Escreva uma função para extrair a parte da hora do tempo variável de entrada
E devolva-o como um número inteiro. Por exemplo:

    1) se a hora for 00, seu código deve retornar 0
    2) se a hora for 01, seu código deve retornar 1
    3) se a hora for 21, seu código deve retornar 21

Por favor, devolva a hora como um número inteiro.


In [10]:
def time_to_hour(time):

    hour = int(time.split(':')[0])
    return hour

assert 3 == time_to_hour('03:12:10')

## Exercicio 2 - Análise dos dados

### *Exercicio 2.1*

Para verificar a relação entre o movimento do metro e o clima, precisaremos complementar os dados do arquivo já baixado com os dados do clima.
Nós complementamos para você este arquivo com os dados de clima de Nova Iorque  e disponibilizamos na área de materiais do projeto. Você pode acessa-lo pelo link: https://s3.amazonaws.com/content.udacity-data.com/courses/ud359/turnstile_data_master_with_weather.csv

Agora que temos nossos dados em um arquivo csv, escreva um código python que leia este arquivo e salve-o em um data frame do pandas. 

Dica: 

Utilize o comando abaixo para ler o arquivo:

```python
pd.read_csv('output_list.txt', sep=",")
```



In [11]:
filename = "turnstile_data_master_with_weather.csv"
weather_df = pandas.read_csv(filename, sep=",")

### *Exercicio 2.2*

Agora crie uma função que calcule a quantidade de dias chuvosos, para isso retorne a contagem do numero de dias onde a coluna *"rain"* é igual a 1.

Dica: Você também pode achar que a interpretação de números como números inteiros ou float pode não 
funcionar inicialmente. Para contornar esta questão, pode ser útil converter
esses números para números inteiros. Isso pode ser feito escrevendo cast (coluna como inteiro).
Então, por exemplo, se queríamos lançar a coluna maxtempi como um número inteiro, nós devemos
escrever algo como cast (maxtempi as integer) = 76, em oposição a simplesmente
onde maxtempi = 76.

In [12]:

def num_rainy_days(df):
    df['rain'] = df['rain'].astype(int)
    num = len(df[df.rain == 1].groupby('DATEn'))
    return num

num_rainy_days(weather_df)

10

### *Exercicio 2.3*

Calcule se estava nebuloso ou não (0 ou 1) e a temperatura máxima para fog (isto é, a temperatura máxima para dias nebulosos).

In [13]:
def max_temp_aggregate_by_fog(df, fog):
    return df[df['fog'] == fog]['maxtempi'].max()
    
nofog = max_temp_aggregate_by_fog(weather_df, 0)
fog = max_temp_aggregate_by_fog(weather_df, 1)

print "fog: ", fog, " no_fog: ", nofog


fog:  81.0  no_fog:  86.0


### *Exercicio 2.4

Calcule agora a média de 'meantempi' nos dias que são sábado ou domingo (finais de semana):

In [14]:
def avg_weekend_temperature(filename):
    df = pandas.read_csv(filename, sep=",")
    df['weekday'] = pandas.to_datetime(df['DATEn']).dt.dayofweek
    meantempi_weekend = df['meantempi'].astype(float)[(df['weekday'] > 4)]
    return meantempi_weekend.mean()

avg_weekend_temperature("turnstile_data_master_with_weather.csv")

65.10066685403307

### *Exercicio 2.5

Calcule a média da temperatura mánima 'mintempi' nos dias chuvosos onde da temperatura mínima foi maior que do 55 graus:

In [15]:
def avg_min_temperature(filename):
    df = pandas.read_csv(filename, sep=",")
    avg_min_temp_rainy = df[(df.rain == 1.0) & (df.mintempi > 55)]['mintempi'].mean()
    return avg_min_temp_rainy

avg_min_temperature("turnstile_data_master_with_weather.csv")

61.238860398860396

### *Exercicio 2.6

Antes de realizar qualquer análise, pode ser útil olhar para os dados que esperamos analisar. Mais especificamente, vamos examinar as entradas por hora em nossos dados do metrô de Nova York para determinar a distribuição dos dados. Estes dados são armazenados na coluna ['ENTRIESn_hourly'].

Trace dois histogramas nos mesmos eixos para mostrar as entradas quando esta chovendo vs quando não está chovendo. 
Abaixo está um exemplo sobre como traçar histogramas com pandas e matplotlib:

```python
Turnstile_weather ['column_to_graph']. Hist ()
```   
    

In [16]:
import numpy as np
import pandas
import matplotlib.pyplot as plt

def entries_histogram(turnstile_weather):
    
    plt.figure()
    turnstile_weather["ENTRIESn_hourly"][turnstile_weather.rain == 0].hist(range=(1000, 8000))
    turnstile_weather["ENTRIESn_hourly"][turnstile_weather.rain == 1].hist(range=(1000, 8000), alpha = 0.8)
    plt.legend(('Not raining','Rainining'))
    plt.xlabel("Entries hourly")
    return plt

entries_histogram(weather_df)

<module 'matplotlib.pyplot' from '/home/hipertrix/.local/lib/python2.7/site-packages/matplotlib/pyplot.pyc'>

### *Exercicio 2.7

Os dados que acabou de plotar que tipo de ditribuição? Existe diferença na distribuição entre dias chuvosos e não chuvosos?

** Resposta **: Representa uma distribuição assimétrica positiva. Exibe de forma clara uma diferença entre as entradas em dias chuvosos e não chuvosos, evidenciando que em dias não chuvosos existe mais entradas.

### *Exercicio 2.8

Construa uma função que que retorne:

1. A média das entradas com chuva
2. A média das entradas sem chuva




In [17]:
import numpy as np

import pandas

def means(turnstile_weather):
    
    with_rain_mean = turnstile_weather["ENTRIESn_hourly"][turnstile_weather.rain == 1].mean()
    without_rain_mean = turnstile_weather["ENTRIESn_hourly"][turnstile_weather.rain == 0].mean()
    with_rain_mean = round(with_rain_mean, 2)
    without_rain_mean = round(without_rain_mean, 2)
    p = None
    return with_rain_mean, without_rain_mean, p # leave this line for the grader


means(weather_df)

(1105.45, 1090.28, None)

Responda as perguntas abaixo de acordo com a saida das suas funções:

1. Qual a média das entradas com chuva?
2. Qual a média das entradas sem chuva?


** Resposta **: 

1. A média das entradas com chuva é 1105.45
2. A média das entradas sem chuva é 1090.28

## Exercicio 3 - Map Reduce

### *Exercicio 3.1*

A entrada para esse exercício e o mesmo arquivo da seção anterior (Exercicio 2). Você pode baixar o arquivo neste link:

 https://s3.amazonaws.com/content.udacity-data.com/courses/ud359/turnstile_data_master_with_weather.csv

Varmos criar um mapeador agora. Para cada linha de entrada, a saída do mapeador deve IMPRIMIR (não retornar) a UNIT como uma chave e o número de ENTRIESn_hourly como o valor. Separe a chave e o valor por uma guia. Por exemplo: 'R002 \t 105105.0'

Exporte seu mapeador em um arquivo chamado mapper_result.txt e envie esse arquivo juntamente com a sua submissão. O código para exportar seu mapeador já está escrito no código abaixo.




In [18]:
import sys

def mapper():
    
    first_line = True
    for line in sys.stdin:
        
        if first_line == True:
            first_line = False
            continue

        data = line.strip().split(',')
        if len(data) < 7:
            continue

        print "{0}\t{1}".format(data[1],data[6])


mapper()
sys.stdin = open('turnstile_data_master_with_weather.csv')
sys.stdout = open('mapper_result.txt', 'w')

### *Exercicio 3.2*

Agora crie o redutor. Dado o resultado do mapeador do exercicio anterior, o redutor deve imprimir(Não retornar) uma linha por UNIT, juntamente com o número total de ENTRIESn_hourly. Ao longo de maio (que é a duração dos nossos dados), separados por uma guia. Um exemplo de linha de saída do redutor pode ser assim: 'R001 \t 500625.0'

Você pode assumir que a entrada para o redutor está ordenada de tal forma que todas as linhas correspondentes a uma unidade particular são agrupados. No entanto a saida do redutor terá repetição pois existem lojas que aparecem em locais diferentes dos arquivos.

Exporte seu redutor em um arquivo chamado reducer_result.txt e envie esse arquivo juntamente com a sua submissão.

In [19]:
import sys

def reducer():
    
    old_unit = None
    total_entries = 0
    
    for line in sys.stdin:
        data = line.strip().split("\t")

        if len(data) != 2:
            continue
        
        unit, entries = data

        if old_unit and old_unit != unit:
            print old_unit, "\t", int(total_entries)
            total_entries = 0

        old_unit = unit
        total_entries += float(entries)

    if old_unit != None:
        print old_unit, "\t", int(total_entries)

reducer()
